src/functions.php

```diff
@@ -377,13 +386,132 @@ function upload(string $xml, $config)
         throw new \Exception('Upload failed');
     }
 }
+
+
+/**
+ * Downloads the old phone book from Fritzbox
+ *
+ * @param   Api   $fritz
+ * @param   array $config
+ * @return  SimpleXMLElement with the old existing phonebook
+ */
+function downloadPhonebook(Api $fritz, array $config)
+{
+    $formfields = [
+        'PhonebookId' => $config['phonebook']['id'],
+        'PhonebookExportName' => $config['phonebook']['name'],
+        'PhonebookExport' => "",
+    ];
+    $result = $fritz->postFile($formfields, []); // send the command to load existing phone book
+    if (substr($result, 0, 5) !== "<?xml") {
+        error_log("ERROR: Could not load old phonebook with ID=".$config['phonebook']['id']);
+        return new SimpleXMLElement("<xml><empty /></xml>");
+    }
+    $xmlPhonebook = simplexml_load_string($result);
+    return $xmlPhonebook;
+}
+
+
+/**
+ * Get quickdial and vanity special attributes from given XML phone book
+ *
+ * @param   SimpleXMLElement    $xmlPhonebook
+ * @return  array               [] or map with {phonenumber@CardDavUID}=>SimpleXMLElement-Attributes
+ */
+function getPhoneNumberAttributes(SimpleXMLElement $xmlPhonebook)
+{
+    if (!property_exists($xmlPhonebook, "phonebook")) {
+        return [];
+    }
+
+    $specialAttributes = [];
+    foreach ($xmlPhonebook->phonebook->contact as $contact) {
+        foreach ($contact->telephony->number as $number) {
+            if ((isset($number->attributes()->quickdial) && $number->attributes()->quickdial >= 0)
+                || (isset($number->attributes()->vanity) && $number->attributes()->vanity != "")) {
+                $key = generateUniqueKey($number, $contact->carddav_uid);
+                $specialAttributes[$key] = $number->attributes();
+            }
+        }
+    }
+    return $specialAttributes;
+}
+
+
+/**
+ * Restore special attributes (quickdial, vanity) in given target phone book
+ *
+ * @param   SimpleXMLElement    $xmlTargetPhoneBook
+ * @param   array               $oldSpecialAttributes [] or map key => attributes
+ * @return  SimpleXMLElement    out phonebook with restored special attributes
+ */
+function mergePhoneNumberAttributes(SimpleXMLElement $xmlTargetPhoneBook, array $oldSpecialAttributes)
+{
+    if (!$oldSpecialAttributes) {
+        return $xmlTargetPhoneBook;
+    }
+    error_log("Restoring old special attributes (quickdial, vanity)".PHP_EOL);
+    foreach ($xmlTargetPhoneBook->phonebook->contact as $contact) {
+        foreach ($contact->telephony->number as $number) {
+            $key = generateUniqueKey($number, $contact->carddav_uid);
+            if (array_key_exists($key, $oldSpecialAttributes)) {
+                foreach (['quickdial','vanity'] as $attribute) {
+                    if (property_exists($oldSpecialAttributes[$key], $attribute)) {
+                        if (!property_exists($number->attributes(), $attribute)) {
```

Comment:
do we need the if case? Can we just assign the attribute in any case as below using ` $number[$attribute]`?

Reply:
OK, PHP can handle this. Thx!